Improvements from eigengrau

eigengrau opened the pull request:
- apply mode name rewriting to pre-existing buffers
- fix fundamental-mode handling

eigengrau committed:
Special-case mode-name rewriting for fundamental mode

Emacs will not run hooks for «after-major-mode-change» for newly-created
fundamental mode buffers. In order to catch these, we now advise
«get-create-buffer».

An alternative would have been to hook into «buffer-list-update». However, this
is also emitted each time windows are selected, buffers killed, so that it is
run much more frequently. Additionally, the hook appears to only be run for the
buffer of the selected window, so that getting the newly created
fundamental-mode buffer would have been more involved.

mrkkrp commented on cyphejor.el:
```diff
@@ -148,9 +156,15 @@ information."
       (advice-add 'wdired-change-to-dired-mode :after #'cyphejor--hook)
-    (advice-remove 'wdired-change-to-dired-mode #'cyphejor--hook)))
+    (advice-remove 'wdired-change-to-dired-mode #'cyphejor--hook))
+  (when cyphejor-mode
```
This may be a problematic feature. What happens when I decide to disable `cyphejor-mode`? The modes affected by this loop will continue to live with changed names (at least from the code alone it seems like that would be the case, I haven't tried).

eigengrau replied:
That is true, but this is unchanged behaviour as far as cyphejor-mode is concerned. I.e., we didn’t store the old major-mode names anyway, so even without these changes the buffers remained live with the rewritten mode names after disabling cyphejor mode.

eigengrau replied:
While playing around with the code I considered also adding a commit which allows restoring mode names of existing buffers, but there are some tricky cases involved and the usecase seemed to not really be worth it (since my guess is most people just enable the mode once and then leave it enabled; at least I do).

eigengrau commented:
Actually I’m not sure bb78bee catches all instances of fundamental-mode buffers popping up. It *does* address the user manually creating a new buffer. However, e.g., I’ve noticed that the mode name for the`*Flycheck error messages*` buffer in the buffer list still pops up unmodified.

mrkkrp commented on cyphejor.el:
```diff
@@ -148,9 +156,15 @@ information."
   (funcall (if cyphejor-mode #'add-hook #'remove-hook)
            'after-change-major-mode-hook
            #'cyphejor--hook)
+  (funcall (if cyphejor-mode #'add-hook #'remove-hook)
```
Can you remind me how to reproduce a situation when the current code fails to handle fundamental? I tried to switch to fundamental after creating a buffer with `M-x fundamental-mode` and it worked (name of mode became just "F"). I remember there should be a case when it'll stay with "Fundamental" highlighter, but I don't remember how to reproduce it.

mrkkrp replied:
It may be better to leave fundamental alone, because it's supposed to be a "vanilla" environment with all minor modes disabled and having many hooks just for this one special mode/using hacks is not a good thing IMO.

eigengrau replied:
The case when fundamental isn’t handled arises when you do `'M-x switch-to-buffer` and enter a buffer name that doesn’t exist yet. I’m not sure there’s something inherently evil in switching the mode name for fundamental mode buffers. It just seems they are tricky to catch.

mrkkrp replied:
Strange, for me it creates a buffer in `text-mode`...

eigengrau replied:
Ah. It looks like this is because the default value of `major-mode` for me is `fundamental-mode` for some reason. I didn’t expressly set this iirc. But you should be able to reproduce the issue after doing `setq-default major-mode 'fundamental-mode)`.

eigengrau replied:
I’d say if it’s not possible to hook into `get-buffer-create` (even though the docs say this should be possible) it’s not worth bothering with the special hook. On the other hand this is somewhat of an OCD-trigger in ibuffer.

eigengrau replied:
Hm, looks like the hook should indeed be [called](https://github.com/emacs-mirror/emacs/blob/99ceefa8ec4f9993663492cfcce6bb82a94569c1/src/buffer.c#L595) from `get-buffer-create`.

eigengrau replied:
Ah, okay. The hook is only run from within either the current buffer or all currently displayed buffers. I’ll check which one it is.

mrkkrp reviewed (commented):
Thanks for you contribution! Here are some comments.

eigengrau commented:
Okay, so, while the docs say that `buffer-list-update` is run by `get-buffer-create` (which flycheck uses), that doesn’t actually seem to be the case. After displaying the buffer, the hook runs. I wonder if Emacs changed behavior there.

eigengrau committed:
Apply mode name rewriting to pre-existing buffers

Depending on where during Emacs initialization cyphejor-mode is engaged, it can
be the case that certain buffers we would like to have renamed already exist.
E.g., this will frequently be the case for the Messages buffer. In order to
remove the need for users to concern themselves with exactly when cyphejor-mode
is loaded, we now run «cyphejor--hook» for all buffers which are live when the
mode is activated.

eigengrau commented:
So, I amended the special-casing for fundamental mode. I’m proposing to advise `get-buffer-create` in order to handle fundamental-mode buffers. This seems to be much less involved than using `buffer-list-update-hook`. Any thoughts?

mrkkrp commented:
I guess I'm OK with advising `get-buffer-create`, what about my second comment?

eigengrau commented:
If you mean clobbering the original mode names, I replied to your comment:

That is true, but this is unchanged behaviour as far as cyphejor-mode is concerned. I.e., we didn’t store the old major-mode names anyway, so even without these changes the buffers remained live with the rewritten mode names after disabling cyphejor mode.

While playing around with the code I considered also adding a commit which allows restoring mode names of existing buffers, but there are some tricky cases involved and the usecase seemed to not really be worth it (since my guess is most people just enable the mode once and then leave it enabled; at least I do).

eigengrau commented:
Something to note (maybe to address later) is that Cyphejor doesn’t rewrite `tags-table-mode` in a buffer created by `M-x visit-tags-table`. This remains true even after advising `get-buffer-create`.

eigengrau commented:
Okay, `tags-table-mode` is just being silly in that it’s not a real mode and it fudges with `mode-name` manually. So this probably isn’t something we could potentially cover.

mrkkrp commented:
OK, let's merge this, thank you!

eigengrau commented:
Thanks for making cyphejor!
